website/files/models/base.py

```diff
@@ -0,0 +1,588 @@
+            index = int(getattr(cls, 'is_file', 2))
+
+            if cls_map[index] is not None:
+                raise ValueError('Conflicting providers')
+
+            cls_map[index] = cls
+
+
+class FileNode(object):
+    """The base class for the entire files storage system.
+    Use for querying on all files and folders in the database
+    """
+    FOLDER, FILE, ANY = 0, 1, 2
+
+    __metaclass__ = FileNodeMeta
+
+    @classmethod
+    def create(cls, **kwargs):
+        """A layer of abstraction around the creation of FileNodes.
+        Provides hook in points for subclasses
+        This is used only for GUID creation.
+        """
+        assert hasattr(cls, 'is_file') and hasattr(cls, 'provider'), 'Must have is_file and provider to call create'
+        kwargs['is_file'] = cls.is_file
+        kwargs['provider'] = cls.provider
+        return cls(**kwargs)
+
+    @classmethod
+    def get_or_create(cls, node, path):
+        """Tries to find a FileNode with node and path
+        See FileNode.create
+        """
+        path = '/' + path.lstrip('/')
+        try:
+            # Note: Possible race condition here
+            # Currently create then find is not super feasable as create would require a
+            # call to save which we choose not to call to avoid filling  the database
+            # with notfound/googlebot files/url. Raising 404 errors may roll back the transaction however
+            return cls.find_one(
+                Q('node', 'eq', node) &
+                Q('path', 'eq', path)
+            )
+        except NoResultsFound:
+            return cls.create(node=node, path=path)
+
+    @classmethod
+    def resolve_class(cls, provider, _type=2):
+        """Resolves a provider and type to the appropriate subclass.
+        Usage:
+            >>> FileNode.resolve_class('box', FileNode.ANY)  # BoxFileNode
+            >>> FileNode.resolve_class('dropbox', FileNode.FILE)  # DropboxFile
+        :rtype: Subclass of FileNode
+        """
+        return PROVIDER_MAP[provider][int(_type)]
+
+    @classmethod
+    def _filter(cls, qs=None):
+        """Creates an odm query to limit the scope of whatever search method
+        to the given class.
+        :param qs RawQuery: An odm query or None
+        :rtype: RawQuery or None
+        """
+        # Build a list of all possible contraints leaving None when appropriate
+        # filter(None, ...) removes all falsey values
+        qs = filter(None, (qs,
+            Q('is_file', 'eq', cls.is_file) if hasattr(cls, 'is_file') else None,
+            Q('provider', 'eq', cls.provider) if hasattr(cls, 'provider') else None,
+        ))
+        # If out list is empty return None; there's no filters to be applied
+        if not qs:
+            return None
+        # Use reduce to & together all our queries. equavilent to:
+        # return q1 & q2 ... & qn
+        return reduce(lambda q1, q2: q1 & q2, qs)
```

Comment:
Minor, but you can easily make this forwards-compatible by using `functools.reduce`.